src/server/pfs/fuse/filesystem.go

```diff
@@ -1,729 +1,216 @@
 package fuse
 
 import (
-	"bytes"
 	"fmt"
-	"io"
 	"os"
+	"os/signal"
 	"path"
-	"path/filepath"
 	"strings"
 	"sync"
-	"syscall"
-	"time"
 
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
 	"github.com/pachyderm/pachyderm/src/client"
-	pfsclient "github.com/pachyderm/pachyderm/src/client/pfs"
-	pfs_server "github.com/pachyderm/pachyderm/src/server/pfs/server"
-
-	"bazil.org/fuse"
-	"bazil.org/fuse/fs"
-	"github.com/gogo/protobuf/types"
-	log "github.com/sirupsen/logrus"
-	"golang.org/x/net/context"
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/codes"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
 )
 
-type filesystem struct {
-	apiClient *client.APIClient
-	Filesystem
-	inodes map[string]uint64
-	lock   sync.RWMutex
-}
-
-func newFilesystem(
-	apiClient *client.APIClient,
-	commitMounts []*CommitMount,
-) *filesystem {
-	return &filesystem{
-		apiClient: apiClient,
-		Filesystem: Filesystem{
-			commitMounts,
-		},
-		inodes: make(map[string]uint64),
+// Mount pfs to mountPoint, opts may be left nil.
+func Mount(c *client.APIClient, mountPoint string, opts *Options) error {
+	nfs := pathfs.NewPathNodeFs(newFileSystem(c, opts.getCommits()), nil)
```

Comment:
It can be, it's actually valid to call functions with a `nil` receiver, since it's effectively just an argument. You only segfault if you try to access fields on it with the function, so `getCommits` is implemented as:

```go
func (o *Options) getCommits() map[string]string {
	if o == nil {
		return nil
	}
	return o.Commits
}
```

Reply:
Ah, I didn't know you could have a nil receiver. Good to know.